File: url-shortener-backend/src/url/url.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { CreateUrlDTO } from './dto/create-url.dto';
import { Url } from './entity/url.entity';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { createShortLink } from 'src/lib/shortener.lib';
import { AnalyticsDTO } from './dto/analytics.dto';
import { AnalyticsData } from './type/analytics.type';
import { getRandomInt } from 'src/lib/number.util';

@Injectable()
export class UrlService {
  constructor(@InjectRepository(Url) private urlRepository: Repository<Url>) {}

  async createUrl(link: string, id: number) {
    const url = this.urlRepository.create({
      url: link,
      short: createShortLink(),
      clicks: 0,
      user: {
        id: id,
      },
    });

    const result = await this.urlRepository.save(url);

    return {
      success: true,
      message: 'Url saved',
      data: result,
    };
  }

  async getUrlById(short: string) {
    const url = await this.urlRepository.findOne({
      where: {
        short: short,
      },
    });

    if (!url) {
      return {
        success: false,
        message: 'Url not found with the short id provided',
        data: null,
      };
    }

    return {
      success: true,
      message: 'url found',
      data: url,
    };
  }

  async getUrls(userId: number) {
    Logger.log(`userid: ${userId}`);
    try {
      const urls = await this.urlRepository.find({
        where: {
          user: {
            id: userId,
          },
        },
      });

      if (!urls) {
        return {
          success: false,
          message: `No url found with provided user id: ${userId}`,
          data: null,
        };
      }

      return {
        success: true,
        message: 'Url list',
        data: urls,
      };
    } catch (e: any) {
      return {
        success: false,
        message: `Error getting urls: ${e.message}`,
        data: null,
      };
    }
  }

  async increaseCount(short: string) {
    const increment = await this.urlRepository.increment(
      {
        short: short,
      },
      'clicks',
      1,
    );

    const update = await this.urlRepository.update(
      {
        short: short,
      },
      {
        updatedAt: new Date(),
      },
    );

    return {
      success: true,
      message: 'Click count updated',
      data: update,
    };
  }

  async getClickAnalytics(userId: number) {
    try {
      const urls = await this.urlRepository.find({
        where: {
          user: {
            id: userId,
          },
        },
      });

      if (!urls || urls.length == 0) {
        return {
          success: true,
          data: {
            clicks: 0,
            urls: 0,
            sales: 0,
            leads: 0,
          },
          message: 'No url found',
          error: null,
        };
      }

      let clicks = 0;
      let urlCount = 0;
      urls.forEach((url) => {
        clicks += url.clicks;
        urlCount++;
      });

      return {
        success: true,
        data: {
          clicks: clicks,
          urls: urlCount,
          sales: 0,
          leads: 0,
        },
        message: 'url and click count data',
        error: null,
      };
    } catch (e) {
      return {
        success: false,
        data: {
          clicks: 0,
          urls: 0,
          sales: 0,
          leads: 0,
        },
        message: `Soemthing went wrong: ${e.message}`,
        error: e.message,
      };
    }
  }

  async getAnalytics(id: number, dto: AnalyticsDTO){
    try {
      const result = await this.urlRepository.find({
        where: {
          user: {
            id: id
          }
        },
        order: {
          updatedAt: 'DESC',
        },
      });

      let total = 0;
      const grouped: Map<string, number> = result.reduce(
        (acc, url) => {
          // Normalize to YYYY-MM-DD
          const dateKey = url.updatedAt.toISOString().split('T')[0];

          if (!acc[dateKey]) {
            acc[dateKey] = 0;
          }
          total += url.clicks;
          acc[dateKey] = acc[dateKey] + url.clicks;
          return acc;
        },
        {} as Map<string, number>,
      );


      
      return {
        success: true,
        message: 'clicks per day',
        data: {
          stats: grouped,
          total_clicks: total,
          leads: getRandomInt(20, 200),
          sales: getRandomInt(26, 223)
        },
      };
    } catch (e) {
      return {
        success: false,
        data: null,
        message: `Soemthing went wrong: ${e.message}`,
        error: e.message,
      };
    }
  }

  // async getAnalytics(id: number, dto: AnalyticsDTO) {
  //   try {
  //     const result = await this.urlRepository
  //       .createQueryBuilder('url')
  //       .select('DATE(url.updatedAt)', 'date')
  //       .addSelect('COUNT(url.clicks)', 'total')
  //       .groupBy('DATE(url.updatedAt)')
  //       .orderBy('date', 'DESC')
  //       .getRawMany();

  //     const r = result.map((row) => ({
  //       date: row.date,
  //       total: Number(row.total)
  //     }));

  //     return {
  //       success: true,
  //       message: 'clicks per day',
  //       data: r,
  //     };
  //   } catch (e) {
  //     return {
  //       success: false,
  //       data: null,
  //       message: `Soemthing went wrong: ${e.message}`,
  //       error: e.message,
  //     };
  //   }
  // }

  // async getAnalytics(id: number, dto: AnalyticsDTO) {
  //   try {
  //     const result = await this.urlRepository.find({
  //       order: {
  //         updatedAt: 'DESC',
  //       },
  //     });

  //     const grouped = result.reduce(
  //       (acc, url) => {
  //         // Normalize to YYYY-MM-DD
  //         const dateKey = url.createdAt.toISOString().split('T')[0];

  //         if (!acc[dateKey]) {
  //           acc[dateKey] = [];
  //         }

  //         acc[dateKey].push(url);
  //         return acc;
  //       },
  //       {} as Record<string, Url[]>,
  //     );

  //     return {
  //       success: true,
  //       message: 'clicks per day',
  //       data: grouped,
  //     };
  //   } catch (e) {
  //     return {
  //       success: false,
  //       data: null,
  //       message: `Soemthing went wrong: ${e.message}`,
  //       error: e.message,
  //     };
  //   }
  // }
}

File: url-shortener-backend-go/service/url_service.go
package service

import (
	"os"

	"github.com/anbioc/url-shortener/url-shortener-backend-go/config"
	applogger "github.com/anbioc/url-shortener/url-shortener-backend-go/logger"
	"github.com/anbioc/url-shortener/url-shortener-backend-go/model"
	"github.com/anbioc/url-shortener/url-shortener-backend-go/utils"
	"gorm.io/gorm"
)

type UrlService struct {
	env    *config.Env
	db     *gorm.DB
	logger *applogger.Logger
}

type AnalyticsDTO struct {
	To   string `json:"to"`
	From string `json:"from"`
}

func NewService(env *config.Env, db *gorm.DB, logger *applogger.Logger) *UrlService {
	return &UrlService{
		env:    env,
		db:     db,
		logger: logger,
	}
}

func (service *UrlService) GetUrlList(userId uint) ([]model.Url, error) {
	var urls []model.Url
	var user model.User

	err := service.db.Preload("Urls").Where("id = ?", userId).First(&user).Error
	urls = user.Urls

	// for _, n := range urls {
	// 	log.Printf("url userid: %d and url: %v", n.UserId, n.Url)
	// }

	return urls, err
}

func (service *UrlService) CreateUrl(url string, userId uint) (model.Url, error) {
	var urlEntity model.Url

	urlEntity.Url = url
	urlEntity.Short, _ = utils.CreateShort()
	urlEntity.UserId = userId
	urlEntity.Clicks = 0

	if os.Getenv("ENV") == "production" {
		if err := service.db.AutoMigrate(&urlEntity); err != nil {

			return urlEntity, err
		}
	}

	if err := service.db.Create(&urlEntity).Error; err != nil {
		return urlEntity, err
	}

	return urlEntity, nil

}

func (service *UrlService) GetUrlByShort(short string) (model.Url, error) {
	var url model.Url

	err := service.db.Preload("User").Where("Short = ?", short).First(&url).Error

	return url, err
}

func (service *UrlService) IncreaseCount(short string) (model.Url, error) {
	var url model.Url

	err := service.db.Model(
		&model.Url{}).Where(
		"Short = ?", short).UpdateColumn(
		"Clicks", gorm.Expr("Clicks + ?", 1)).First(&url).Error

	return url, err
}

func (service *UrlService) GetClickAnalytics(userId uint) (int, int, error) {

	var user model.User

	err := service.db.Preload("Urls").Where("id = ?", userId).First(&user).Error

	if err != nil {
		return 0, 0, err
	}

	if len(user.Urls) == 0 {
		return 0, 0, err
	}

	var clicks = 0
	var urlCount = 0

	for _, item := range user.Urls {
		clicks += int(item.Clicks)
		urlCount++

	}

	return clicks, urlCount, nil

}

func (service *UrlService) GetAnalytics(userId uint, analyticsDTO *AnalyticsDTO) (map[string]int, int, error) {
	// var user model.User
	var urls []model.Url

	err := service.db.Where(`"userId" = ?`, userId).Order(`"createdAt" DESC`).Find(&urls).Error

	if err != nil {
		return map[string]int{}, 0, err
	}

	acc := make(map[string]int)
	total := 0

	for _, url := range urls {
		// dateKey := strings.Split(url.UpdatedAt.Format(time.RFC3339), "T")[0]
		dateKey := url.UpdatedAt.Format("2006-03-02")
		if _, ok := acc[dateKey]; !ok {
			acc[dateKey] = 0
		}

		acc[dateKey] = acc[dateKey] + int(url.Clicks)

		total += int(url.Clicks)

	}

	return acc, total, nil

}
